### PR TITLE
domain: report min start timestamp of TiDB server

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
The server current min start timestamp could be reported to PD, it may be used to calculate a more reasonable safe point for GC.

### What is changed and how it works?
Report min start timestamp to PD.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
1. start a TiDB cluster.
1. `ETCDCTL_API=3 ./etcdctl --endpoints="http://172.16.5.34:2380" get / --prefix` gets the key `/tidb/server/minstartts/xxx`
1. start a connection and begin a transaction, then get the keys from PD again.
1. commit the transaction, then get the keys from PD again.

Code changes

 - Has exported function/method change

### CODE REVIEW

File: domain/info.go [hunk 1]
```diff
@@ -144,6 +158,47 @@ func (is *InfoSyncer) RemoveServerInfo() {
...
 }
 
+// storeMinStartTS stores self server min start timestamp to etcd.
+func (is *InfoSyncer) storeMinStartTS(ctx context.Context) error {
+	if is.etcdCli == nil {
+		return nil
+	}
+	return util.PutKVToEtcd(ctx, is.etcdCli, keyOpDefaultRetryCnt, is.minStartTSPath,
+		strconv.FormatUint(is.minStartTS, 10),
+		clientv3.WithLease(is.session.Lease()))
```
Comment (MyonKeminta) [inline domain/info.go]: Does all TiDB servers updates the same key in etcd? If so, how do TiDB servers know whether there's another TiDB server who has a smaller `minStartTs`?

Comment (MyonKeminta) [reply domain/info.go]: Oh I got it. LGTM.

File: domain/info.go [hunk 2]
```diff
@@ -144,6 +158,47 @@ func (is *InfoSyncer) RemoveServerInfo() {
...
+
+// UpdateMinStartTS updates self server min start timestamp from etcd.
+func (is *InfoSyncer) UpdateMinStartTS() {
+	if is.manager == nil {
+		// Server may not start in time.
+		return
+	}
+	pl := is.manager.ShowProcessList()
+	var minStartTS uint64 = math.MaxUint64
+	for _, info := range pl {
```
Comment (coocood) [inline domain/info.go]: What if `len(pl) == 0`?

**Comment by coocood** [comment]
If a tidb-server is killed -9, how long will the minStartTS be kept in PD?

**Comment by jackysp** [comment]
> If a tidb-server is killed -9, how long will the minStartTS be kept in PD?

1 min. It used the same TTL as server info, https://github.com/pingcap/tidb/blob/fbc6a84630cdcaec45ded9faf1b8e244c44c02bd/domain/info.go#L221

File: domain/info.go (hunk 2 above)
Comment (jackysp) [reply domain/info.go]: Report max uint64 to PD.

Comment (jackysp) [reply domain/info.go]: Report max uint64 to PD.

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

File: domain/info.go [hunk 3]
```diff
@@ -49,6 +54,9 @@ type InfoSyncer struct {
 	etcdCli        *clientv3.Client
 	info           *ServerInfo
 	serverInfoPath string
+	minStartTS     uint64
```
Comment (zimulala) [inline domain/info.go]: It seems that we don't need this field.

Comment (zimulala) [reply domain/info.go]: It will be used by the next PR.

**Pull request merged**